src/bitfield_index.hpp

```diff
@@ -0,0 +1,63 @@
+// Copyright 2020 Chia Network Inc
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//    http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+
+struct bitfield_index
+{
+    // cache the number of set bits evey n bits. This is n. For a bitfield of
+    // size 2^32, this means a 2 MiB index
+    static inline const int64_t kIndexBucket = 16 * 1024;
+
+    bitfield_index(std::vector<bool> const& b) : bitfield_(b)
+    {
+        uint64_t counter = 0;
+        auto it = bitfield_.begin();
+        index_.reserve(bitfield_.size() / kIndexBucket);
```

Comment:
`index_.reserve(bitfield_.size() / kIndexBucket);`

Should this be `index_.reserve((bitfield_.size() / kIndexBucket)+1);`, or is the index not used for the last bucket where bucket size < kIndexBucket ?

Reply:
no, I think this is right. This provides an index for the number of set bits at the start of every `kIndexBucket` bits. So it rounds down.